.github/workflows/publish-to-test-pypi.yml

```diff
@@ -32,12 +37,17 @@ jobs:
         .
 
     - name: Publish distribution 📦 to Test PyPI
+      if: github.event_name == 'push'
       uses: pypa/gh-action-pypi-publish@release/v1
       with:
         password: ${{ secrets.TEST_PYPI }}
         repository-url: https://test.pypi.org/legacy/
+        skip_existing: true
+
     - name: Publish distribution 📦 to PyPI
-      if: startsWith(github.ref, 'refs/tags')
+#      if: startsWith(github.ref, 'refs/tags')
+      # The line above is the old code from PyPA, the line below is from @lwasser
+      if: github.event_name == 'release'
```

Comment:
this looks good @jsdodge ! you could use skip_existing on the real pypi section of the build as well if you want it to not push an existing tagged distribution.

also i think you can delete distributions from pypi - but many would say that is a bad idea :) but i've i think done that when just trying to get things to work to avoid file buildup.

do you have any questions on this? i really don't have much to add

Reply:
Thanks @lwasser ! I do have questions about the triggering events. I'm confused by the appearance of the lines below, since I understand them to restrict the triggering event to the main branch. Do they?

https://github.com/dodge-research-group/thztools/blob/d7e627851ae0198bf8dbe1b368c87ed9f66f7420/.github/workflows/publish-to-test-pypi.yml#L5-L10

Reply:
great question @jsdodge 

This section below tells github actions to trigger an event when you push to main.  You could add more branches if you wish. or you could tell it to trigger on all branches with a * . but what you have there restricts the action to be triggered by a push to main ONLY. you are correct
```
   push: 
     branches: 
       - main 
 ```

however with github actions you are also tell it to trigger on other events.
So to push to real pypi, we add what is below which triggers on a published release (not a draft a published release). 
```
   release: 
     types: [published] 
````

So in its entirety the `on:` statement above says 
"only run this action when there is a published release OR a push to main branch. 

does that help?

Reply:
Ah-hah: I read those statements to have AND between them, not OR. That answers my question, thanks!